frontend/app/src/components/StatusWidget/StatusWidget.tsx

```diff
@@ -378,128 +355,153 @@ class StatusWidget extends PureComponent<StatusWidgetProps, State> {
+          </StyledConnectionStatusLabel>
+        </StyledConnectionStatus>
+      </Tooltip>
     )
-  }
+  }, [connectionState, statusMinimized])
 
-  private static getConnectionStateUI(
-    state: ConnectionState
-  ): ConnectionStateUI | undefined {
-    switch (state) {
-      case ConnectionState.INITIAL:
-      case ConnectionState.PINGING_SERVER:
-      case ConnectionState.CONNECTING:
-        return {
-          icon: Ellipses,
-          label: "Connecting",
-          tooltip: "Connecting to Streamlit server",
-        }
+  const renderWidget = useCallback((): ReactNode => {
+    if (isConnected()) {
+      if (
+        scriptRunState === ScriptRunState.RUNNING ||
+        scriptRunState === ScriptRunState.RERUN_REQUESTED
+      ) {
+        // Show scriptIsRunning when the script is actually running,
+        // but also when the user has just requested a re-run.
+        // In the latter case, the server should get around to actually
+        // re-running the script in a second or two, but we can appear
+        // more responsive by claiming it's started immediately.
+        return renderScriptIsRunning()
+      }
+      if (scriptChangedOnDisk) {
+        return renderRerunScriptPrompt()
+      }
+    }
 
-      case ConnectionState.CONNECTED:
-        return undefined
+    return renderConnectionStatus()
+  }, [
+    isConnected,
+    renderConnectionStatus,
+    renderRerunScriptPrompt,
+    renderScriptIsRunning,
+    scriptChangedOnDisk,
+    scriptRunState,
+  ])
+
+  // The StatusWidget fades in on appear and fades out on disappear.
+  // We keep track of our most recent result from `renderWidget`,
+  // via `this.curView`, so that we can fade out our previous state
+  // if `renderWidget` returns null after returning a non-null value.
+  const curView = useRef<ReactNode>()
+  const prevView = curView.current
+  curView.current = renderWidget()
+
+  if (isNullOrUndefined(curView.current) && isNullOrUndefined(prevView)) {
+    return <></>
+  }
 
-      case ConnectionState.DISCONNECTED_FOREVER:
-      default:
-        return {
-          icon: Warning,
-          label: "Error",
-          tooltip: "Unable to connect to Streamlit server",
-        }
-    }
+  let animateIn: boolean
+  let renderView: ReactNode
+  if (notNullOrUndefined(curView.current)) {
+    animateIn = true
+    renderView = curView.current
+  } else {
+    animateIn = false
+    renderView = prevView
   }
+
+  // NB: the `timeout` value here must match the transition
+  // times specified in the StatusWidget-*-active CSS classes
+  return (
+    <CSSTransition
+      appear={true}
+      in={animateIn}
+      timeout={200}
+      unmountOnExit={true}
+      classNames="StatusWidget"
+    >
+      <StyledStatusWidget
+        key="StatusWidget"
+        className="stStatusWidget"
+        data-testid="stStatusWidget"
+      >
+        {renderView}
+      </StyledStatusWidget>
+    </CSSTransition>
+  )
 }
 
 export default withTheme(StatusWidget)
```

Comment:
**suggestion:** `withTheme` is no longer necessary since this uses the `useTheme` hook

Reply:
[43ea2cb](https://github.com/streamlit/streamlit/pull/9851/commits/43ea2cba4292ccdfa38b50f7a4e2247cb54d6243)